Refactor ingestion code & store push when ingesting task

armenzg opened the pull request:
This makes the main function a little shorter by moving chunks of code
into their own functions.

When ingesting a task we will ingest the push if missing.

armenzg commented on treeherder/etl/management/commands/ingest.py:
```diff
@@ -149,10 +219,16 @@ async def await_futures(fs):
-    acquire_connection()
     logger.info("Loading into DB:\t%s", pulse_job["taskId"])
-    JobLoader().process_job(pulse_job, root_url)
-    release_connection()
```
This block is the only logic change.

klahnakoski commented on treeherder/etl/management/commands/ingest.py:
```diff
@@ -61,7 +61,77 @@ def release_connection():
 
-async def handleTaskId(taskId, root_url):
+def ingest_pr(pr_url, root_url):
+    splitUrl = pr_url.split("/")
+    org = splitUrl[3]
+    repo = splitUrl[4]
```
I suggest 

    _, _, _, org, repo, _, pull_number = pr_url.split("/")

klahnakoski replied:
...assuming I know how deep that directory path is

klahnakoski commented on treeherder/etl/management/commands/ingest.py:
```diff
@@ -372,86 +448,31 @@ def add_arguments(self, parser):
         )
 
     def handle(self, *args, **options):
+        INGESTION_TYPES = [
+            'task',
+            'pr',
+            'git-push',
+            'git-pushes',
+            'push',
+        ]
```
Use an enum, or named constants.  `INGESTION_TYPES[0]` is a hard to understand

klahnakoski commented on treeherder/etl/management/commands/ingest.py:
```diff
@@ -149,10 +219,16 @@ async def await_futures(fs):
-    release_connection()
+    try:
+        acquire_connection()
+        JobLoader().process_job(pulse_job, root_url)
```
Should this be in `finally`?  Since it is run here and in the exception.

klahnakoski commented on treeherder/etl/management/commands/ingest.py:
```diff
@@ -149,10 +219,16 @@ async def await_futures(fs):
+    except MissingPushException:
+        logger.warning('The push was not in the DB. We are going to try that first')
+        _ingest_push(pulse_job["origin"]["project"], pulse_job["origin"]["revision"])
+        JobLoader().process_job(pulse_job, root_url)
```
...or maybe this is a retry?

armenzg replied:
It is a retry.

klahnakoski commented on treeherder/etl/management/commands/ingest.py:
```diff
@@ -149,10 +219,16 @@ async def await_futures(fs):
-    JobLoader().process_job(pulse_job, root_url)
-    release_connection()
+    try:
+        acquire_connection()
```
replace 

```python

def acquire_connection():
    """
    Decrement database resource count. If resource count is 0, block thread
    until resource count becomes 1 before decrementing again.
    """
    conn_sem.acquire()


def release_connection():
    """
    Close thread's conneciton to database and increment database resource count.
    """
    connection.close()
    conn_sem.release()

```

with  

```python
class Connection(object):

    def __enter__(self):
        conn_sem.acquire()
        
    def __exit__(self, exc_type, exc_val, exc_tb):
        connection.close()
        conn_sem.release()
```

then you may 

```python

with Connection():
    try:
        JobLoader().process_job(pulse_job, root_url)
    except MissingPushException:
        logger.warning('The push was not in the DB. We are going to try that first')
        _ingest_push(pulse_job["origin"]["project"], pulse_job["origin"]["revision"])
        JobLoader().process_job(pulse_job, root_url)

klahnakoski replied:
... one other thing, 

```python
        connection.close()
        conn_sem.release()
```

I am a bit concerned that the connection is closed, but apparently never opened.  I assume this works despite the lack of symmetry.

armenzg commented:
I did something odd and I've opened a new PR in #6666

I addressed your concerns and will land it after passing the tests.
Thanks for the Connection trick!

armenzg commented:
Landed via #6666.
